test/test_unary_ufuncs.py

```diff
@@ -112,31 +134,42 @@ def generate_numeric_tensors(device, dtype, *,
         vals = _unsigned_int_vals
     else:  # dtypes is a signed integer type
         assert dtype in (torch.int8, torch.int16, torch.int32, torch.int64)
-        large_vals = _large_int_vals if include_large_values else tuple()
-        vals = _int_vals + large_vals
+        vals = _int_vals
 
-    assert len(vals) < medium_length
+    return generate_tensors_from_vals(vals, device, dtype, domain)
 
-    # Constructs the large tensor containing vals
-    large_tensor = make_tensor(large_size, device=device, dtype=dtype, low=domain[0], high=domain[1])
 
-    # Inserts the vals at an odd place
-    large_tensor[57][offset:offset + len(vals)] = torch.tensor(vals, device=device, dtype=dtype)
+def generate_numeric_tensors_hard(device, dtype, *,
+                                  domain=(None, None)):
+    is_signed_integral = dtype in (torch.int8, torch.int16, torch.int32, torch.int64)
+    assert dtype.is_floating_point or dtype.is_complex or is_signed_integral
 
-    # Takes a medium sized copy of the large tensor containing vals
-    medium_tensor = large_tensor[57][offset:offset + medium_length]
+    if dtype.is_floating_point:
+        vals = _large_float_vals
+    elif dtype.is_complex:
+        vals = tuple(complex(x, y) for x, y in chain(product(_large_float_vals, _large_float_vals),
+                                                     product(_float_vals, _large_float_vals),
+                                                     product(_large_float_vals, _float_vals)))
+    else:
+        vals = _large_int_vals
 
-    # Constructs small tensors (4 elements)
-    small_tensors = (t for t in torch.split(medium_tensor, 4))
+    return generate_tensors_from_vals(vals, device, dtype, domain)
 
-    # Constructs scalar tensors
-    scalar_tensors = (t.squeeze() for t in torch.split(medium_tensor, 1))
 
-    # Tensors with no elements
-    empty_sizes = ((0,), (0, 3, 3), (1, 0, 5), (6, 0, 0, 0), (3, 0, 1, 0))
-    empty_tensors = (torch.empty(size, device=device, dtype=dtype) for size in empty_sizes)
+def generate_numeric_tensors_extremal(device, dtype, *,
+                                      domain=(None, None)):
+    assert dtype.is_floating_point or dtype.is_complex
+
+    vals = []
+    if dtype.is_floating_point:
+        vals = _float_extremals
+    elif dtype.is_complex:
+        vals = tuple(complex(x, y) for x, y in chain(product(_float_extremals, _float_extremals),
+                                                     product(_float_vals, _float_extremals),
+                                                     product(_float_extremals, _float_vals)))
+
+    return generate_tensors_from_vals(vals, device, dtype, domain)
```

Comment:
This is OK for now, but I would prefer this test generate the extremal vals and just make one scalar tensor for each of them, plus take a large tensor that has NaNs, -infs, and infs scattered throughout it instead of in one concentrated section. A simple helper to "scatter" these values could be used. A density of 1/20 elements could also be used for each of them in the large tensor.